File: src/main/kotlin/com/ederfmatos/reactive/java/MonoOne.java
package com.ederfmatos.reactive.java;

import reactor.core.publisher.Mono;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public class MonoOne<T> implements One<T> {

    private final Mono<T> value;

    public MonoOne(Mono<T> value) {
        this.value = value;
    }

    @Override
    public <R> One<R> map(Function<T, R> transformer) {
        return new MonoOne<>(value.map(transformer));
    }

    @Override
    public <R> One<R> flatMap(Function<T, One<R>> transformer) {
        return new MonoOne<>(value.flatMap(t -> oneToMono(transformer.apply(t))));
    }

    @Override
    public <R> One<R> then(Supplier<One<R>> transformer) {
        return new MonoOne<>(value.then(Mono.defer(() -> oneToMono(transformer.get()))));
    }

    @Override
    public <R> One<R> thenConsume(Supplier<R> transformer) {
        return new MonoOne<>(value.then(Mono.defer(() -> Mono.justOrEmpty(transformer.get()))));
    }

    @Override
    public One<T> ifEmpty(Supplier<One<T>> supplier) {
        return new MonoOne(value.switchIfEmpty(Mono.defer(() -> oneToMono(supplier.get()))));
    }

    @Override
    public One<T> ifEmptyReturn(Supplier<T> supplier) {
        return new MonoOne(value.switchIfEmpty(Mono.fromSupplier(supplier)));
    }

    @Override
    public One<T> filter(Predicate<T> filter) {
        return new MonoOne<>(value.filter(filter));
    }

    @Override
    public One<T> filterNot(Predicate<T> filter) {
        return new MonoOne<>(value.filter(filter.negate()));
    }

    @Override
    public One<T> onNext(Consumer<T> consumer) {
        return new MonoOne(value.doOnNext(consumer));
    }

    @Override
    public One<T> onError(Consumer<Throwable> consumer) {
        return new MonoOne(value.doOnError(consumer));
    }

    @Override
    public One<T> onEmpty(Runnable runnable) {
        return new MonoOne(value.switchIfEmpty(Mono.defer(() -> {
            runnable.run();
            return Mono.empty();
        })));
    }

    @Override
    public One<T> onSuccess(Consumer<T> consumer) {
        return new MonoOne(value.doOnSuccess(consumer));
    }

    @Override
    public One<T> onComplete(Runnable runnable) {
        return new MonoOne(value.doOnTerminate(runnable));
    }

    private <R> Mono<R> oneToMono(One<R> one) {
        return ((MonoOne<R>) one).value;
    }

    public static <T> One<T> from(T value) {
        return new MonoOne<>(Mono.justOrEmpty(value));
    }

    public static <T> One<T> empty() {
        return new MonoOne<>(Mono.empty());
    }

}

File: src/main/kotlin/com/ederfmatos/reactive/kotlin/MonoOne.kt
package com.ederfmatos.reactive.kotlin

import reactor.core.CoreSubscriber
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.switchIfEmpty

class MonoOne<T : Any>(private val value: Mono<T>) : Mono<T>(), One<T> {
    override fun <R : Any> map(block: (T) -> R): One<R> {
        return value.map(block).let(::MonoOne)
    }

    override fun <R : Any> flatMap(block: (T) -> One<R>): One<R> {
        return MonoOne(value.flatMap { (block.invoke(it) as MonoOne<R>).value })
    }

    override fun filter(block: (T) -> Boolean): One<T> {
        return MonoOne(value.filter(block))
    }

    override fun filterNot(block: (T) -> Boolean): One<T> {
        return MonoOne(value.filter { !block.invoke(it) })
    }

    override fun onNext(block: (T) -> Unit): One<T> {
        return MonoOne(value.doOnNext(block))
    }

    override fun ifEmpty(block: () -> One<T>): One<T> {
        return MonoOne(value.switchIfEmpty {
            (block() as MonoOne<T>).value
        })
    }

    override fun ifEmptyReturn(block: () -> T): One<T> {
        return MonoOne(value.switchIfEmpty {
            just(block())
        })
    }

    override fun onError(block: (Throwable) -> Unit): One<T> {
        return MonoOne(value.doOnError(block))
    }

    override fun onEmpty(block: () -> Unit): One<T> {
        return value.switchIfEmpty(defer {
            block()
            Mono.empty()
        }).let(::MonoOne)
    }

    override fun onComplete(block: () -> Unit): One<T> {
        return value.doOnSuccess { block() }.let(::MonoOne)
    }

    override fun onSuccess(block: (T) -> Unit): One<T> {
        return value.doOnSuccess(block).let(::MonoOne)
    }

    override fun subscribe(actual: CoreSubscriber<in T>) {
        value.subscribe(actual)
    }

    companion object {
        fun <T : Any> from(value: T): One<T> = MonoOne(just(value))
        fun <T : Any> empty(): One<T> = MonoOne(Mono.empty())
    }
}